lib/symboldatabase.cpp

```diff
@@ -7453,7 +7453,8 @@ void SymbolDatabase::setValueTypeInTokenList(bool reportDebugWarnings, Token *to
                     else if (type <= ValueType::Type::LONG && !MathLib::isDec(tokStr) && mSettings.platform.isLongValue(value >> 2)) {
                         type = ValueType::Type::LONG;
                         sign = ValueType::Sign::UNSIGNED;
-                    } else if (mSettings.platform.isLongLongValue(unsignedSuffix ? (value >> 1) : value))
+                    } else if (mSettings.platform.isLongLongValue(unsignedSuffix ? (value >> 1) : value) ||
```

Comment:
The `cppcheck --debug --verbose 12110.c` output:
```
##AST
[12110.c:2]
(
|-if
`-/ 'unsigned long long'
  |-$-9223372036854775808 'unsigned long long'
  `--1 'signed int'
```
It would seem to me the that -9223372036854775808 should be a  `'signed long long'`. maybe that is exactly what you accomplish with the fix but it's not clear to me.

Reply:
Yes, the absolute value of `LLONG_MIN` exceeds the maximum value of `long long` (checked in `platform.h`), so it defaults to unsigned. But maybe the function in `platform.h` should work differently? I agree that this is not very clear.

Reply:
hmm.. no I was mistaken the type of -9223372036854775808 should be `'unsigned long long'`. I trust Clang:
```
void f() {
    auto x = 0x8000000000000000;
    auto y = -9223372036854775808;
}
```
Output from clang:
```
$ clang -cc1 -ast-dump 12110.cpp
.......
    |-DeclStmt 0x32193800 <line:2:5, col:32>
    | `-VarDecl 0x32193660 <col:5, col:14> col:10 x 'unsigned long':'unsigned long' cinit
    |   `-IntegerLiteral 0x321936c8 <col:14> 'unsigned long' 9223372036854775808
    |-DeclStmt 0x321939e0 <line:3:5, col:34>
    | `-VarDecl 0x32193858 <col:5, col:15> col:10 y 'unsigned long long':'unsigned long long' cinit
    |   `-UnaryOperator 0x321938e0 <col:14, col:15> 'unsigned long long' prefix '-'
    |     `-IntegerLiteral 0x321938c0 <col:15> 'unsigned long long' 9223372036854775808
```
It seems that the value 9223372036854775808 is a 'unsigned long long' if you write it in decimal format but a 'unsigned long' if you write it in hexadecimal format. That is probably described in the specification.

Reply:
btw this clang output was written on a platform where sizeof(long) is 8 and sizeof(long long) is 8.

Reply:
why the type can be different for the same value in decimal and hexadecimal format is described here:

https://stackoverflow.com/questions/12499592/c-hex-constant-type